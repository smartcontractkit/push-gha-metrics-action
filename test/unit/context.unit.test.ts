import type { context } from "@actions/github";
import { DeepMockProxy, mockDeep, MockProxy } from "jest-mock-extended";
import * as fixtures from "../fixtures/dynamic-name";
import { writeFileSync } from "node:fs";
import { join } from "node:path";
import {
  fetchWorkflowRunContext,
  fetchJobRunContext,
  getGithubContext,
  fetchContext,
} from "../../src/context";
import type { ContextOverrides, Octokit } from "../../src/context.types";

function mockResponse(data: any): any {
  return {
    headers: {},
    status: "200" as const,
    data,
    url: "",
  };
}

describe("Context", () => {
  let mockContext: MockProxy<typeof context>;
  let mockClient: DeepMockProxy<Octokit>;
  let mockContextOverrides: ContextOverrides;
  const env = JSON.parse(JSON.stringify(process.env));

  describe("Event", () => {
    describe("Pull Requests", () => {
      beforeEach(() => {
        process.env = env;
        mockClient = mockDeep<Octokit>();
      });

      describe.each(fixtures.pullRequest.fixtures)("%# Job Attempts", (f) => {
        beforeEach(() => {
          // make the ended at timestamp be 30 seconds after the one of the jobs in the current workflow started
          const estimatedEndedAtUnixSeconds =
            Math.floor(
              Date.parse(
                f.api.listJobsForWorkflowRunAttempt.jobs[0].started_at,
              ) / 1000,
            ) + 30;

          mockContextOverrides = {
            estimatedEndedAtUnixSeconds,
            jobName: "generate-fixtures-name-1",
          };

          mockClient.rest.actions.getWorkflowRunAttempt.mockResolvedValue(
            mockResponse(getWorkflowRunAttempt),
          );
          mockClient.rest.actions.listJobsForWorkflowRunAttempt.mockResolvedValue(
            mockResponse(listJobsForWorkflowRunAttempt),
          );
        });

        const { getWorkflowRunAttempt, listJobsForWorkflowRunAttempt } = f.api;
        mockContext = {
          ...f.context,
          // The fixture is missing the getters since its plain json
          // So we add our own to test if we're properly creating a JSON obj based off of
          // https://github.com/actions/toolkit/blob/main/packages/github/src/context.ts
          get issue(): { owner: string; repo: string; number: number } {
            return { owner: "mockOwner", repo: "mockRepo", number: 1 };
          },
          get repo(): { owner: string; repo: string } {
            return { owner: "mockOwner", repo: "mockRepo" };
          },
        };

        describe(getGithubContext.name, () => {
          it("should handle required env vars", () => {
            delete process.env.GITHUB_EVENT_PATH;
            delete process.env.GITHUB_RUN_ATTEMPT;

            expect(() =>
              getGithubContext(mockContext),
            ).toThrowErrorMatchingInlineSnapshot(
              `"GITHUB_EVENT_PATH must exist to obtain context"`,
            );

            process.env.GITHUB_EVENT_PATH = "./my-event-path";

            expect(() =>
              getGithubContext(mockContext),
            ).toThrowErrorMatchingInlineSnapshot(
              `"GITHUB_RUN_ATTEMPT must exist to get workflow run attempt"`,
            );

            process.env.GITHUB_RUN_ATTEMPT = "1";

            expect(() => getGithubContext(mockContext)).not.toThrowError();
          });

          it("should return a correctly hydrated context", () => {
            process.env.GITHUB_RUN_ATTEMPT = "1";
            process.env.GITHUB_EVENT_PATH = "./my-event-path";
            const result = getGithubContext(mockContext, {});
            // Have to serialize / parse otherwise snapshotting fails
            // most likely due to how we mock this object
            expect(JSON.parse(JSON.stringify(result))).toMatchSnapshot();
          });
        });

        describe(fetchJobRunContext.name, () => {
          it("should error out when a job lookup fails", async () => {
            process.env.GITHUB_RUN_ATTEMPT = "1";
            process.env.GITHUB_EVENT_PATH = "./my-event-path";
            const githubContext = getGithubContext(mockContext);
            expect.assertions(1);

            await expect(() =>
              fetchJobRunContext(mockClient, githubContext),
            ).rejects.toMatchSnapshot();
          });

          it("should work when a matching job name is found", async () => {
            process.env.GITHUB_RUN_ATTEMPT = "1";
            process.env.GITHUB_EVENT_PATH = "./my-event-path";
            // jobName was grabbed by looking through the fixtures
            const githubContext = getGithubContext(
              mockContext,
              mockContextOverrides,
            );

            const result = await fetchJobRunContext(
              mockClient,
              githubContext,
              mockContextOverrides,
            );
            expect(result).toMatchSnapshot();
          });
        });

        describe("fetchWorkflowRunContext", () => {
          it("should error when run_started_at is null", async () => {
            mockClient.rest.actions.getWorkflowRunAttempt.mockResolvedValueOnce(
              mockResponse({
                ...getWorkflowRunAttempt,
                run_started_at: null,
              }),
            );
            expect.assertions(1);
            const githubContext = getGithubContext(mockContext);
            await expect(
              fetchWorkflowRunContext(mockClient, githubContext),
            ).rejects.toMatchSnapshot();
          });

          it("should work", async () => {
            const githubContext = getGithubContext(mockContext);
            const workflowRunContext = await fetchWorkflowRunContext(
              mockClient,
              githubContext,
            );

            expect(workflowRunContext).toMatchSnapshot();
          });
        });

        describe("fetchContext", () => {
          it("should work", async () => {
            const res = await fetchContext(
              mockClient,
              mockContext,
              mockContextOverrides,
            );

            const stringifiedRes = JSON.stringify(res);
            expect(JSON.parse(stringifiedRes)).toMatchSnapshot();

            /**
             * This data is used in other unit tests as a mock, this makes it easy
             * to update them
             */
            if (process.env.UPDATE_CONTEXT_MOCK) {
              const snapshotPath = "__mocks__";
              writeFileSync(
                join(__dirname, snapshotPath, `fetchContext${f.index}.json`),
                stringifiedRes,
              );
            }
          });
        });
      });
    });
  });
});
