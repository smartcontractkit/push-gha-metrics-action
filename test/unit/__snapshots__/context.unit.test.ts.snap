// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context Event Pull Requests 0 Job Attempts fetchContext should work 1`] = `
Object {
  "event": Object {
    "actor": "HenryNguyen5",
    "eventName": "pull_request",
    "ref": "refs/pull/21/merge",
    "repo": Object {
      "owner": "mockOwner",
      "repo": "mockRepo",
    },
    "sha": "512d1a168cf626018888b44dc35cdd6e79b9df8a",
  },
  "jobRun": Object {
    "estimatedEndedAtUnixSeconds": 1655254904,
    "id": 6891428337,
    "jobName": "generate-fixtures-name-1",
    "name": "generate-fixtures-name-1",
    "startedAt": "2022-06-15T01:01:13Z",
    "startedAtUnixSeconds": 1655254873,
    "url": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/jobs/6891428337",
  },
  "workflowRun": Object {
    "createdAt": "2022-06-15T01:01:00Z",
    "createdAtUnixSeconds": 1655254860,
    "runAttempt": 1,
    "runId": 2499110858,
    "runNumber": 28,
    "runStartedAt": "2022-06-15T01:01:00Z",
    "runStartedAtUnixSeconds": 1655254860,
    "updatedAt": "2022-06-15T01:01:17Z",
    "updatedAtUnixSeconds": 1655254877,
    "url": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/runs/2499110858",
    "workflowName": "Push Metrics CI",
    "workflowUrl": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/workflows/28123629",
  },
}
`;

exports[`Context Event Pull Requests 0 Job Attempts fetchJobRunContext should error out when a job lookup fails 1`] = `
[Error: No job for job name: "generate-fixtures" found during self-lookup, invalid job name given?
      Available jobs names + ids: my dummy job|6891428272,generate-fixtures-name-1|6891428337,generate-fixtures-name-2|6891428403
      ]
`;

exports[`Context Event Pull Requests 0 Job Attempts fetchJobRunContext should work when a matching job name is found 1`] = `
Object {
  "estimatedEndedAtUnixSeconds": 1655254904,
  "id": 6891428337,
  "name": "generate-fixtures-name-1",
  "startedAt": "2022-06-15T01:01:13Z",
  "startedAtUnixSeconds": 1655254873,
  "url": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/jobs/6891428337",
}
`;

exports[`Context Event Pull Requests 0 Job Attempts fetchWorkflowRunContext should error when run_started_at is null 1`] = `[Error: Could not find "run_started_at" for workflow run attempt runId:2499110858|attemptNumber:1, github api down / inconsistent?]`;

exports[`Context Event Pull Requests 0 Job Attempts fetchWorkflowRunContext should work 1`] = `
Object {
  "createdAt": "2022-06-15T01:01:00Z",
  "createdAtUnixSeconds": 1655254860,
  "runStartedAt": "2022-06-15T01:01:00Z",
  "runStartedAtUnixSeconds": 1655254860,
  "updatedAt": "2022-06-15T01:01:17Z",
  "updatedAtUnixSeconds": 1655254877,
  "url": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/runs/2499110858",
  "workflowUrl": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/workflows/28123629",
}
`;

exports[`Context Event Pull Requests 0 Job Attempts getGithubContext should return a correctly hydrated context 1`] = `
Object {
  "actor": "HenryNguyen5",
  "eventName": "pull_request",
  "jobName": "generate-fixtures",
  "ref": "refs/pull/21/merge",
  "repo": Object {
    "owner": "mockOwner",
    "repo": "mockRepo",
  },
  "runAttempt": 1,
  "runId": 2499110858,
  "runNumber": 28,
  "sha": "512d1a168cf626018888b44dc35cdd6e79b9df8a",
  "workflowName": "Push Metrics CI",
}
`;

exports[`Context Event Pull Requests 1 Job Attempts fetchContext should work 1`] = `
Object {
  "event": Object {
    "actor": "HenryNguyen5",
    "eventName": "pull_request",
    "ref": "refs/pull/21/merge",
    "repo": Object {
      "owner": "mockOwner",
      "repo": "mockRepo",
    },
    "sha": "512d1a168cf626018888b44dc35cdd6e79b9df8a",
  },
  "jobRun": Object {
    "estimatedEndedAtUnixSeconds": 1655500786,
    "id": 6942977656,
    "jobName": "generate-fixtures-name-1",
    "name": "generate-fixtures-name-1",
    "startedAt": "2022-06-17T21:19:17Z",
    "startedAtUnixSeconds": 1655500757,
    "url": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/jobs/6942977656",
  },
  "workflowRun": Object {
    "createdAt": "2022-06-17T21:19:08Z",
    "createdAtUnixSeconds": 1655500748,
    "runAttempt": 1,
    "runId": 2499110858,
    "runNumber": 28,
    "runStartedAt": "2022-06-17T21:19:07Z",
    "runStartedAtUnixSeconds": 1655500747,
    "updatedAt": "2022-06-17T21:19:19Z",
    "updatedAtUnixSeconds": 1655500759,
    "url": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/runs/2499110858",
    "workflowName": "Push Metrics CI",
    "workflowUrl": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/workflows/28123629",
  },
}
`;

exports[`Context Event Pull Requests 1 Job Attempts fetchJobRunContext should error out when a job lookup fails 1`] = `
[Error: No job for job name: "generate-fixtures" found during self-lookup, invalid job name given?
      Available jobs names + ids: my dummy job|6942977557,generate-fixtures-name-1|6942977656,generate-fixtures-name-2|6942977735
      ]
`;

exports[`Context Event Pull Requests 1 Job Attempts fetchJobRunContext should work when a matching job name is found 1`] = `
Object {
  "estimatedEndedAtUnixSeconds": 1655500786,
  "id": 6942977656,
  "name": "generate-fixtures-name-1",
  "startedAt": "2022-06-17T21:19:17Z",
  "startedAtUnixSeconds": 1655500757,
  "url": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/jobs/6942977656",
}
`;

exports[`Context Event Pull Requests 1 Job Attempts fetchWorkflowRunContext should error when run_started_at is null 1`] = `[Error: Could not find "run_started_at" for workflow run attempt runId:2499110858|attemptNumber:1, github api down / inconsistent?]`;

exports[`Context Event Pull Requests 1 Job Attempts fetchWorkflowRunContext should work 1`] = `
Object {
  "createdAt": "2022-06-17T21:19:08Z",
  "createdAtUnixSeconds": 1655500748,
  "runStartedAt": "2022-06-17T21:19:07Z",
  "runStartedAtUnixSeconds": 1655500747,
  "updatedAt": "2022-06-17T21:19:19Z",
  "updatedAtUnixSeconds": 1655500759,
  "url": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/runs/2499110858",
  "workflowUrl": "https://api.github.com/repos/smartcontractkit/push-gha-metrics-action/actions/workflows/28123629",
}
`;

exports[`Context Event Pull Requests 1 Job Attempts getGithubContext should return a correctly hydrated context 1`] = `
Object {
  "actor": "HenryNguyen5",
  "eventName": "pull_request",
  "jobName": "generate-fixtures",
  "ref": "refs/pull/21/merge",
  "repo": Object {
    "owner": "mockOwner",
    "repo": "mockRepo",
  },
  "runAttempt": 1,
  "runId": 2499110858,
  "runNumber": 28,
  "sha": "512d1a168cf626018888b44dc35cdd6e79b9df8a",
  "workflowName": "Push Metrics CI",
}
`;
